futures-executor/src/spawn.rs

```diff
@@ -0,0 +1,137 @@
+use futures_core::{Future, Async, Poll};
+use futures_core::never::Never;
+use futures_core::task::{self, Context};
+use futures_channel::oneshot::{channel, Sender, Receiver};
+use futures_util::FutureExt;
+
+use std::thread;
+use std::sync::Arc;
+use std::sync::atomic::Ordering;
+use std::panic::{self, AssertUnwindSafe};
+use std::sync::atomic::AtomicBool;
+
+/// dox
+#[derive(Debug)]
+pub struct Spawn<F>(Option<F>);
+
+/// TODO: dox
+pub fn spawn<F>(f: F) -> Spawn<F>
+    where F: Future<Item = (), Error = ()> + 'static + Send
+{
+    Spawn(Some(f))
+}
+
+impl<F: Future<Item = (), Error = ()> + Send + 'static> Future for Spawn<F> {
+    type Item = ();
+    type Error = Never;
+    fn poll(&mut self, cx: &mut Context) -> Poll<(), Never> {
+        cx.spawn(self.0.take().unwrap());
+        Ok(Async::Ready(()))
+    }
+}
+
+/// dox
+#[derive(Debug)]
+pub struct SpawnWithHandle<F>(Option<F>);
+
+/// TODO: dox
+pub fn spawn_with_handle<F>(f: F) -> SpawnWithHandle<F>
```

Comment:
Can we also offer `spawn_with_handle` on `task::Context`?

Reply:
Problem is, the implementation relies on a bunch of the other crates. We could of course offer an extension trait here and put it in the prelude, but didn't seem worth it right now.